### PR TITLE
Added DS record type

### PR DESCRIPTION
**Pull request by kochie**
Added support for Delegation Signer record. Should partially address #31 and fully address #75. I've added support for Cloudflare at the moment and I'm hoping to add support for all compatible providers soon.

Current providers with DS enabled
- [x] Cloudflare
- [ ] Gandi
- [x] dnsimple

Hoping to get some feedback and verification that the data types are compatible with the [RFC](https://www.rfc-archive.org/getrfc.php?rfc=4034). 

Thanks!

### CODE REVIEW

**Comment by captncraig** [comment]
Wow, thanks for working on this. I'm a bit confused about what situations you would be specifying the DS records directly though.

Looking at the cloudflare console, it looks like the default workflow is to take the DS record they have created and use for signing, and pass that to the upstream registrar. Most other providers I have looked at seem to have this kind of model where you "activate" it, then need to register the DS records upstream. You would never have access to the private keys that the provider uses is this case.

In what situation would you manually need to specify the DS records? Are you manually signing records yourself? Are you adding DS records from another provider for consistency? 

I feel like I'm missing something, since it is definitely something you can do in their web console, but I'm not clear why.

My imagined design for dnssec has been each compatible provider implements some `GetDSRecords` api call, and compatible registrars would have a corresponding `SetDSRecords` process. But last I checked (a while ago admittedly), none of the registrars I use had such an api. I imagine some may now.

**Comment by captncraig** [comment]
Ok, after writing a little code I can see how it may be good to specify DS records directly, in case you need to manually enter them or a provider doesn't have a read api or something.

DS seem analagous to how we handle NS records perhaps. We try to automatically infer which nameservers to use from your providers and register them appropriately, but you can also directly specify nameserver records if you need something special. DS may need a little magic and a little flexibility. 

Please share how you envision using all this as well.

File: integrationTest/integration_test.go [hunk 1]
```diff
@@ -601,6 +610,15 @@ func makeTests(t *testing.T) []*TestCase {
 		)
 	}
 
+	// DS
+	if !providers.ProviderHasCabability(*providerToRun, providers.CanUseRoute53Alias) {
```
Comment (captncraig) [inline integrationTest/integration_test.go]: This looks like it was started and not finished. I'd love to have some integration tests to implement it for more providers.

File: pkg/js/parse_tests/024-ds.js [hunk 2]
```diff
@@ -0,0 +1,4 @@
+D("foo.com","none",
+    DS("foo.com", 1000, 13, 2, "AABBCCDDEEFF"),
```
Comment (captncraig) [inline pkg/js/parse_tests/024-ds.js]: We should use `@` instead of `foo.com`, even in tests.

File: pkg/normalize/validate.go [hunk 3]
```diff
@@ -302,7 +304,7 @@ func NormalizeAndValidateConfig(config *models.DNSConfig) (errs []error) {
 			}
 
 			// Canonicalize Targets.
-			if rec.Type == "CNAME" || rec.Type == "MX" || rec.Type == "NAPTR" || rec.Type == "NS" || rec.Type == "SRV" {
+			if rec.Type == "CNAME" || rec.Type == "DS" || rec.Type == "MX" || rec.Type == "NAPTR" || rec.Type == "NS" || rec.Type == "SRV" {
```
Comment (captncraig) [inline pkg/normalize/validate.go]: I don't think ds should use this case. There is no hostname in the target fields.

**Comment by kochie** [comment]
Hey @captncraig, thanks for the feedback, I'll update the PR with the fixes. 

Just thought I'd give my perspective on adding the record. Your 100% right in that I rarely need to specify a DS record manually, however there is one use case which will probably start to become more commonplace, and that's transferring domains to new registrars. 

Basically I had a few domains with Namecheap that I transferred over to Cloudflare which automatically scanned and added a lot of junk records (>500) that Namecheap had included, since I also had DNSSEC enabled it copied over the DS record, which causes dnscontrol to panic when it read the record.

Basically my motivation for this PR was to allow dnscontrol to delete unneeded DS records that are imported when transferring domains using an external system.

💯 agree as well on your idea for how DNSSEC should be applied with dnscontrol, it would be great to simply assign a flag or something when defining a new domain in the config.

File: pkg/normalize/validate.go (hunk 3 above)
Comment (kochie) [reply pkg/normalize/validate.go]: Cool, can you think of any other validation checks for DS?

**Comment by tlimoncelli** [comment]
Please rebase (`git pull --rebase --ff-only --prune`) and push. Thanks!

File: pkg/normalize/validate.go (hunk 3 above)
Comment (tlimoncelli) [reply pkg/normalize/validate.go]: I can't think of any.  We generally don't check the individual fields for fear of locking us into specific constants, etc.

**Comment by patschi** [comment]
Any updates on this @kochie?

I would like to have DS support as well, specially for OVH in my case. So once this is merged I can start looking into getting this added for OVH too. :)

**Comment by tlimoncelli** [comment]
Please rebase and run "go generate". Thanks!

**Comment by tlimoncelli** [comment]
Please rebase and re-run "go generate".  I'll merge this as soon after. 

Thanks and sorry for the delay!

**Comment by kochie** [comment]
It's still giving a gogen error, is this expected?

**Comment by tlimoncelli** [comment]
Did you commit the file that gogen updated?

```
$ go version
go version go1.12.4 darwin/amd64
$ go generate
$ git status
On branch kochie-feature/ds-record-type
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   docs/_includes/matrix.html

no changes added to commit (use "git add" and/or "git commit -a")
$
```

It looks like you need to run `git commit docs/_includes/matrix.html` then `git push`

**Comment by tlimoncelli** [comment]
Closing due to age.  Please re-open if you would like to continue.

**Comment by McNetic** [comment]
I rebased and updated the original pull at https://github.com/StackExchange/dnscontrol/pull/746

I'm not a go developer and did not review the original work besides from what was necessary to get it to work. I was however able to create DS records with the desec provider.

**Comment by tlimoncelli** [comment]
Thanks!  I'll take a look this week.
